crates/l2/docs/state_diffs.md

```diff
@@ -0,0 +1,60 @@
+- The bytecode of every contract deployed on the network.
+- All withdrawal Logs.
+
+After executing each L2 block, the EVM will return the following data:
+
+- A list of every storage slot modified in the block, with their previous and next values. A storage slot is a mapping `(address, slot) -> value`. Note that, on a block, there could be repeated writes to the same slot. In that case, we keep only the latest write; all the others are discarded since they are not needed for state reconstruction.
+- The bytecode of every newly deployed contract. Every contract deployed is then a pair `(address, bytecode)`.
+- A list of withdrawal logs (as explained in milestone 1 we already collect these and publish a merkle root of their values as calldata, but we still need to send them as the state diff).
+- A list of triples `(address, nonce_increase, balance)` for every modified account. The `nonce_increase` is a value that says by how much the nonce of the account was increased on the block (this could be more than one as there can be multiple transactions for the account on the block). The balance is just the new balance value for the account.
+
+The full state diff sent on every block will then be a sequence of bytes encoded as follows. We use the notation `un` for a sequence of `n` bits, so `u16` is a 16-bit sequence and `u96` a 96-bit one, we don’t really care about signedness here; if we don’t specify it, the value is of variable length and a field before it specifies it.
+
+- The first byte is a `u8`: the version header. For now it should always be zero, but we reserve it for future changes to the encoding/compression format.
+- Next come the `ModifiedAccounts` list. The first two bytes (`u16`) are the amount of element it has, followed by its entries. Each entry correspond to an altered address and has the form:
+  - The first byte is the `type` of the modification. The value is a `u8`, constrained to the range `[1; 23]`, computed by adding the following values:
+    - `1` if the balance of the EOA/contract was modified.
+    - `2` if the nonce of the EOA/contract was modified.
+    - `4` if the storage of the contract was modified.
+    - `8` if the contract was created and the bytecode is previously unknown.
+    - `16` if the contract was created and the bytecode is previously known.
+  - The next 20 bytes, a `u160`, is the address of the modified account.
+  - If the balance was modified (i.e. `type & 0x01 == 1`), the next 32 bytes, a `u256`, is the new balance of the account.
+  - If the nonce was modified (i.e. `type & 0x02 == 2`), the next 2 bytes, a `u16`, is the increase in the nonce.
+  - If the storage was modified (i.e. `type & 0x04 == 4`), the next 2 bytes, a `u16`, is the number of storage slots modified. Then come the sequence of `(key_u256, new_value_u256)` key value pairs with the modified slots.
```

Comment:
This can be improved. Assuming the common case is modify less than 128 slots, the length encoding could be:
- If length is `[1; 127]`, it only uses one byte (`[0x01; 0x7f]`)
- If length is `[128; 32895]`, it uses two bytes and is encoded as `0x7F80 + length` (`[0x8000; 0xFFFF]`)

Note: As every slot uses 64 bytes (`key_u256` + `value_256`), having 32895 slots modified implies more than 2MB, way more than a EIP-4844 blob size. Taking this in account, we can even tweak the border between 1 byte and 2 bytes encoding to maximize the 1 byte encoding use (I think this would be around 248)

Reply:
Going deep with the improvement, using 248 (0xF8) as the border, we have that:
- If length is `[1; 247]`, it only uses one byte (`[0x01; 0xF7]`)
- If length is `[248;2295]`, it only uses two bytes and is encoded as `0xF708 + length` (`[0xF800; 0xFFFF]`)
- If length is `> 2295` it doesn't fit in the blob